### ISSUE TITLE
Unexpected behaviour using pitch fly

### ISSUE THREAD

**Issue by ClobberXD** [minetest/minetest#7946]
##### Issue type

 - Bug report
 
##### Summary

Pitch-fly (327bad2) modified all player movement to be calculated with respect to player's (camera's) pitch, but this introduced some unexpected behaviour: player ascent and descent is also propagated with respect to pitch, so if one was to look as high up (or low down) as possible, ascent/descent would actually be horizontal movement.

With the introduction of pitch fly, I believe the function of separate ascend/descend movement is to allow the player to go up and down without having to look in that direction.

Expected behaviour: Player should go straight up/down irrespective of the pitch.

**Comment by random-geek** [comment]
I agree the non-vertical aspect can get annoying at times, but it could be considered a feature. I assume this is probably too trivial a difference to make a setting for it?

**Comment by paramat** [comment]
> Player should go straight up/down irrespective of the pitch.

I agree, and disagree with a setting for it.

**Comment by paramat** [comment]
Not sure now. Ascend/descend being vertical becomes somewhat redundant in this mode.

**Comment by ClobberXD** [comment]
> Not sure now. Ascend/descend being vertical becomes somewhat redundant in this mode.

Yes. They become meaningless if the resulting movement could even be fully horizontal. Hence my request to make these movements independent of player pitch.

**Comment by random-geek** [comment]
What would happen if the player was looking downward while pressing W and Space simultaneously?

**Comment by ClobberXD** [comment]
The key that was pressed last would take precedence. Try pressing space and the descend key while swimming or with `free_move`. If the last-pressed key is released, the other key controls the movement. That's how it currently works. I agree that this is confusing too, but not as much as moving horizontally while pressing Space.

**Comment by ClobberXD** [comment]
Setting sounds like a good idea, but there'd be too many settings for one trivial behaviour. On the other hand, having to press `S` or `E` to go up in certain cases is the very unintuitive, and makes it really confusing for the younger (read majority) players out there, who probably don't understand the concept of vector dot products :)

**Comment by paramat** [comment]
Whatever is decided i'm :-1: for it to be an option, let's just choose the best behaviour.

**Comment by jastevenson303** [comment]
I can see uses for both methods.  If I am a flying humanoid I want up to go
up.  If I'm a jetplane things change and going straight up makes little
sense.

**Comment by paramat** [comment]
Finally i have tried out pitch fly mode.

The current behaviour or 'up' and 'down' is unfamiliar, but that's expected, it's new. I think we have to be careful to not interpret unfamiliar as 'wrong'.
Current code makes 'up' and 'down' useful at all pitches, if changed they will become a duplicate motion when facing near straight up/down.
Motion working relative to pitch is consistent with forward, back, left and right controls.

So i'm :-1: for the issue suggestion for the short term. Let's at least allow players to get used to the new mode first, as i expect the current behaviour will then be valued more.

**Comment by jastevenson303** [comment]
> while swimming

ClobberXD raises an important point.  If pitch `fly` mode is enabled, then even when you're not in `fly` mode, you raise upwards.  This means you cannot backpedal properly in this mode.

This mode is kinda not correct.

**Comment by ClobberXD** [comment]
Thanks for reviving this issue, jas ;)

@paramat 

> The current behaviour or 'up' and 'down' is unfamiliar, but that's expected, it's new. I think we have to be careful to not interpret unfamiliar as 'wrong'.

This is just so wrong, and has nothing to do with unfamiliarity. One would press the jump key, expecting to go up, not move perpendicular to pitch.

> Let's at least allow players to get used to the new mode first

I suggest we fix this _before_ players get used to pitch-fly, otherwise they'll get accustomed to the wrong movement controls altogether, making it even more difficult to tweak the controls again.

**Comment by paramat** [comment]
Well, i'm correct to write it's unfamiliar behaviour, and correct that 'unfamiliar' cannot be equated to 'wrong', because new things are always unfamiliar.
You're starting with the assumption that Y axis motion is expected by most based on your own expectation.
Also, what is 'expected' is not necessarily the best thing to do, or what players will eventually prefer.
Deciding based on expectation alone prevents progress or anything new happening, because people can't expect new things.

I admit that i find it a little weird at first, but i'm aware this is just due to what i'm used to, maybe it's better this way and has advantages? Maybe i should get used to it?
Essentially, we have to decide on the actual advantages and disadvantages. So do you have any real arguments? You have only mentioned 'expectation'.

The commit message is "In pitch fly mode, you fly to the exact direction you are pointing at, using the forward key. Other move directions are also pitched accordingly." so the behaviour is intented by the author and was merged with that behaviour.
>  otherwise they'll get accustomed to the wrong movement controls

Also a non-argument because you're starting with the assumption that it's wrong without stating any actual disadvantages. "it's wrong because it's unexpected", well, it's obviously unexpected because it's new.

**Comment by jastevenson303** [comment]
That is an awful lot of text.

The point is, if pitch fly mode is enabled and you're in water, you cannot go backwards.  Even if fly mode is off.  That's broken behavior!

**Comment by ClobberXD** [comment]
> You're starting with the assumption that Y axis motion is expected by most based on your own expectation.

Yes, and I assume most other players would expect the same of our movement controls. To decide upon whether the current movement behaviour is correct or not, is the sole purpose of this issue :)

> So do you have any real arguments?

Let's say I'm lighting up a massive cavern with torches while on pitch-fly mode. I would also make use of `A`, `D`, `Space` and `E` to complement my mouse movement, essentially fine-tuning my (pitch-dependent) forward movement. But while `A` and `D` allow me to move left and right irrespective of pitch (as intended), the ascending and descending movements are relative to pitch. If I'm not flying parallel to the ground, ascending and descending will also make me move in the opposite direction of my yaw. e.g. If I'm flying at a bearing of 90 degrees (E) at an incline of 60 degrees, pressing `Space` will create a push in a direction perpendicular to my current motion, essentially resulting in an incline of 105 degrees, and a bearing of 270 degrees (W).

TL;DR: When flying non-parallel to ground in pitch-fly mode, ascending/descending makes me move backwards. I'd consider that broken behaviour. :)

**Comment by paramat** [comment]
However, it's quite likely i will agree with the suggestion, it just seems too soon to decide before 5.0.0 and we are in feature freeze, this isn't a bugfix but an implementation choice.
As a dev i fly around a lot so will test both ways. 5.1.0 will likely be soon after 5.0.0 so if we deal with this then it will only be a few months.
So to clarify :-1: for merging this for 5.0.0, but certainly something to consider in 5.1.0-dev. I'm currently neutral on which implementation seems generally best for MT players.
Until then it seems best to stay with the author's intention and what was merged.

**Comment by ClobberXD** [comment]
While I would like to see this in 5.0.0 itself, I understand that 5.0.0 has already been delayed long enough, and that we're already in the feature freeze (although I'd consider this a bugfix :P). So, 5.1.0 would be okay with me, thanks.

**Comment by jastevenson303** [comment]
https://forum.minetest.net/viewtopic.php?f=18&t=22078&p=343516#p343507

Emphasis on the _in water_ part.

**Comment by ClobberXD** [comment]
To quote:
> Also idk if its a bug or feature but when I press space on water I don't only go upwards but forward as well, and with shift I go move backward and down, that happens when flying too its so annoying.

@paramat :)

**Comment by CasimirKaPazi** [comment]
I have another reason to add for why this is a bug.
In water, when you press forward you move forward, when you look down and press jump, you also move forward.
Combining those, looking down in a 45° angle lets you move faster than normally allowed. Therefor it's a bug that can be used by those few people who find out to get an advantage over the other players.

It will be just the same as we had with sneak elevators. People using a strange bug for their advantage. It's risky to include this in 5.0.0

Other than that, it also limits a players game play. You might want to swim through the ocean, skimming the seafloor for resources. However, when you look down you also move down. But actually you want to move forward. With current controls there is no way to do this.
With the controls as Clobber suggested you could do this and it would also feel very intuitive. It would be similar to how you move at land, you can look down to the floor or into the sky and still move horizontally.

Just try it out play for some time: You are in the water, look up and press jump. What direction would someone new to the game expect? What happens now is that you move backwards. It's counterintuitive.

In my opinion this is a bug that should be fixed before 5.0.0, otherwise it would be 1. a bad start for the new version jump, 2. something ugly people will get used to.

**Comment by paramat** [comment]
> But while A and D allow me to move left and right irrespective of pitch (as intended), the ascending and descending movements are relative to pitch.

True but obvious as a result if geometry.
The intention of the merged commit was movement relative to view direction and the resulting behaviour works as intended.
"the ascending and descending movements are relative to pitch." ... as intended.
So nothing is broken, it's simply a case of preference.

> The point is, if pitch fly mode is enabled and you're in water, you cannot go backwards. Even if fly mode is off. That's broken behavior!

Just tested in water, pitch move enabled, fly mode off, you can go backwards, press S. What do you mean here?

I've seen the comments in the forum thread, that's 4 out of thousands of forum members. Complainers are far more likely to comment than those who are quietly content. So 4 complainers isn't necessarily rconclusive, but we should of course reconsider, and i think it's quite likely the behaviour will be changed (i'm neutral), but too late now.
> In water, when you press forward you move forward, when you look down and press jump, you also move forward.
Combining those, looking down in a 45° angle lets you move faster than normally allowed. Therefor it's a bug that can be used by those few people who find out to get an advantage over the other players.

Pressing W together with A or D probably has the same effect, in or out of water. This diagonal speed advantage is likely not new.
> Other than that, it also limits a players game play. You might want to swim through the ocean, skimming the seafloor for resources. However, when you look down you also move down. But actually you want to move forward. With current controls there is no way to do this.
With the controls as Clobber suggested you could do this and it would also feel very intuitive. It would be similar to how you move at land, you can look down to the floor or into the sky and still move horizontally.

Wut? Just turn pitch move off (tested).
> You are in the water, look up and press jump. What direction would someone new to the game expect? What happens now is that you move backwards. It's counterintuitive.

You would expect up of course, but that already happens with the default standard movement. Pitch move is not standard movement and is therefore likely to result in non-standard movement.

**Comment by CasimirKaPazi** [comment]
> Pitch move is not standard movement

Oh, okay. I got the impression that it is standard now. In that case I don't care.

**Comment by DS-Minetest** [comment]
>Pressing W together with A or D probably has the same effect, in or out of water. This diagonal speed advantage is likely not new.

You are wrong. It is new.
The horizontal velocity is set to the length of the player speed. See [here](https://github.com/minetest/minetest/blob/ad0f20835cd4f64f5d47f537125e9e3157717beb/src/client/localplayer.cpp#L666-L672).

However, I don't see a big problem here. Swimming faster with this method has disadvantages in sight and attacking other players is harder. Furthermore swimming on the surface with holding the jump button might be even faster. And using this method needs skills by the player.

<details><summary>(how I tested this)</summary>

I had made flowing liquids give the player a velocity. See https://github.com/DS-Minetest/minetest/tree/fluidwash. Then I placed a water source and swam towards the source. With pitchfly I was able to swim nearer to the source. Using left and forward didn't change anything.
</details>

**Comment by ClobberXD** [comment]
> no-one's written a PR to change it yet

If that's the issue, I'd be more than happy to do it myself, once I get back on the 6th. :)

**Comment by paramat** [comment]
I am supporting this because, even though i am not sure, i still find current behaviour a little weird even after using it for so long.

**Comment by ClobberXD** [comment]
Thanks, will make a PR soon.

**Comment by ClobberXD** [comment]
Like everything else in life, looking at the code, I get a feeling that this is easier said than done... :)
